docs/design/provisioning.md

```diff
@@ -0,0 +1,223 @@
+
+One configuration is shown below.  The remainder can be found [here](provisioning-examples.json).
+
+```
+  {
+    "kind": "ConfigMap",
+    "apiVersion": "v1",
+    "metadata": {
+      "name": "gold-east",
+      "namespace": "kubernetes.io/provisioners",
+      "labels": {
+        "type": "provisioner-config",
+        "storage-class": "gold",
+        "zone": "us-east"
+      }
+    },
+    "data": {
+      "plugin-name": "kubernetes.io/gce-pd",  
+      "parameters": "{'volumeType':'ssd', 'zone':'us-east'}"  // OPAQUE JSON MAP OF PARAMS AS DEFINED BY THE PLUGIN
+    }
+  }
+  
+```
+
+End users request specific storage by using a PersistentVolumeSelector on their claim.  This example shows a request for "gold" storage in the east zone.
+
+```
+{
+  "kind": "PersistentVolumeClaim",
+  "apiVersion": "v1",
+  "metadata": {
+    "name": "claim2",
+    "namespace": "end-user-namespace"
+  },
+  "spec": {
+    "accessModes": [
+      "ReadWriteOnce"
+    ],
+    "persistentVolumeSelector":{
+    	"matchLabels":{
+    		"storage-class":"gold",
+    		"zone":"us-east",
+    	}
+    }
+    "resources": {
+      "requests": {
+        "storage": "3Gi"
+      }
+    }
+  }
+}
+	
+```
+
+## Labelling PersistentVolumes
+
+A `PersistentVolumeSelector` on claim has the additional benefit of being able to bind to volumes that are manually created as opposed to dynamically through a provisioner.
+
+The same claim above (gold+east) can match on a PersistentVolume that is labelled similarly.
+
+```
+apiVersion: v1
+  kind: PersistentVolume
+  metadata:
+    name: forMyClaim
+    labels:
+    	storage-class: gold
+    	zone: us-east
+  spec:
+    capacity:
+      storage: 15Gi
+    accessModes:
+      - ReadWriteOnce
+    persistentVolumeReclaimPolicy: Recycle
+    nfs:
+      path: /tmp
+      server: 172.17.0.2
+```
+
+
+## Controller behavior
+
+`PersistentVolumeSelector` is being slightly overloaded in what it selects.  The binding controller would use the same selector on two different but related objects:  PersistentVolumes and Provisioners that create PersistentVolumes.
+
+The binder attempts to fulfill a claim in the following order:
+
+* If `pvc.Spec.PersistentVolumeSelector` is non-nil, attempt to find an available PV with a matching label set.
+* If `pvc.Spec.PersistentVolumeSelector` is non-nil and matches a provisioner's labels, create a PV using the provisioner
+* If `pvc.Spec.PersistentVolumeSelector` is nil, attempt to match an available PV by AccessModes and Capacity (current behavior).
+
+Claims remain Pending indefinitely if a match is not found (with or without a selector).  The claim will be bound when a volume or provisioner is created that matches the claim.
+
+
+## Design Considerations
+
+`StorageClass` has been proposed as a field on PersistentVolumeClaim.  An equivalent of that field is currently stewing as an annotation in the current version of dynamic provisioning.
```

Comment:
Let's leave it an annotation for now. It's easier to add things to the API later than remove them.